### DISCUSSION TITLE
Is it possible to handle errors?

### DISCUSSION THREAD

**Discussion by andriykrefer** [nushell/nushell#5830] [Q&A]
Scripts immediately crash on error. Could not find any way to handle errors.
For instance, if ffmpeg is not installed:

```
Error: nu::shell::external_command (link)

  × External command
    ╭─[./build.nu:51:1]
 51 │     
 52 │     if ((ffmpeg -version | complete | get exit_code) != 0) {
    ·          ───┬──
    ·             ╰── can't run executable
 53 │         echo "Installing ffmpeg"
    ╰────
  help: No such file or directory (os error 2)
```

The script crashes and I can not handle the error.

I am sure there are other ways to solve this specific problem, but my point is that I could not find a way to gracefully handle an error before the program exits.

**Comment by hustcer** [comment]
workaround:
```shell
  # Check if some command available in current shell
  def 'is-installed' [ app: string ] {
    ((which $app | length) > 0)
  }

  if (is-installed 'ffmpeg') {
    # Do something here
  } else {
    $'(ansi r)Command `ffmpeg` could not be found, Please install it and try again!(ansi reset)'
    exit --now
  }
```

**Reply by andriykrefer** [reply]
Appreciate the answer. Personally, a added "^env" before ffmpeg as a workaround. But the point of the question is how to handle errors in nushell.

**Comment by kubouch** [comment]
You can use `do -i { ... }` and check for `$nothing`. The error handling needs to be improved, though, there isn't a good way to handle them right now.

**Reply by andriykrefer** [reply]
Thanks, this is a useful workaround that I will try in my scripts

**Comment by kendfss** [comment]
fwiw, this isn't exactly built-in-friendly, but there's also `$env.LAST_EXIT_CODE`. Which you can check as an int in comparison statements.
You can combine that with these, though most people/tools today violate these left right and center, for added granularity.
https://tldp.org/LDP/abs/html/exitcodes.html
https://pastebin.com/dLn8Lk6e (/usr/include/sysexits.h)

**Reply by andriykrefer** [reply]
It does not apply to the case I showed, because the script will throw an error and exit before I am able to check the return code of the (inexistent) external command.

**Comment by merelymyself** [comment]
New functionality, do -c allows you to capture errors. You can then check if it’s an error through describe.
